Simple TensorFlow example. using volume mount and configuration parameter to switch models (#52)
lightbend/cloudflow: examples/tensorflow-akka/README.md

## `tensorflow-akka`

### Problem Definition

A simple pipeline that scores the quality of wines using a TensorFlow model.

> The TensorFlow models must be provided through a persistent volume claim. 

Reviewer: An interesting alternative would be to use [Alpakka with Google Cloud Storage](https://doc.akka.io/docs/alpakka/current/google-cloud-storage.html)

Author: That is an interesting alternative indeed. It might be easier.. I'll have a closer look, thanks @skonto !

The benefit of doing what is done here is that it is a very simple solution combining the volume mount and config features. It also works immediately with runLocal, using local file system. But it does require the PV stuff...

Author: I think we could look at this from a separate example.. leaving it for now.

A `load-models` script is provided to load TensorFlow SavedModelBundles to the volume that is mounted on the model-server streamlet.

The model-server can be configured to use any of the loaded models, by choosing a different SavedModelBundle directory.
(This directory has to be copied first to the volume, before the application is deployed or re-configured.) 

### Required configuration

`model-server.models`
Persistent volume claim for the directory that contains TensorFlow SavedModelBundle directories.

`model-server.model`
The relative directory name under /models that must be used to load the TensorFlow SavedModelBundle from. 

### Generating data
The WineRecordGenerator Streamlet generates wine records.

##### Run
TODO

### Example Deployment on GKE

> * Make sure you have installed a GKE cluster with Cloudflow running as per the [installation guide](https://github.com/lightbend/cloudflow-installer).

Reviewer 2: ```suggestion
* Make sure you have created a GKE cluster and installed Cloudflow as per the [installation guide](https://github.com/lightbend/cloudflow-installer).
```

> Make sure you have access to your cluster:

Reviewer 2: ```suggestion
Verify access to your cluster:
```

...
```

* Install the required PVC for the models.

First create the application namespace (since a PVC is namespaced).

```
$ kubectl create ns tensorflow-akka
```

claim.yaml:

```
apiVersion: v1
kind: PersistentVolumeClaim
metadata:
  name: claim1
  namespace: tensorflow-akka
spec:
  accessModes:
    - ReadWriteMany
  storageClassName: nfs-client
  resources:
    requests:
      storage: 1Gi
```

```
$ kubectl apply -f claim.yaml
```

Copy the models to the persistent volume:

```
$ ./load-models.sh
```
* Deploy the app.

TODO update to use passwd-stdin

```
$ kubectl cloudflow  deploy -u oauth2accesstoken --volume-mount model-server.wine-models=claim1  eu.gcr.io/<projectID>/tensorflow-akka:<version> -p "$(gcloud auth print-access-token)"
```

* Verify that the application is deployed.

```
$ kubectl cloudflow list

NAME              NAMESPACE         VERSION           CREATION-TIME     
tensorflow-akka   tensorflow-akka   8-2a0f65d-dirty   2019-11-25 13:22:44 +0200 EET
```

* Check all pods are running.

```
$ kubectl get pods -n tensorflow-akka
NAME                                                  READY   STATUS    RESTARTS   AGE

TODO

```

> * Verify the application output

Reviewer: @RayRoestenburg it would be useful to add output here for scoring so when users try it to know if they are getting something meaningful or some installation step failed. Probably its part of the TODO above.

Author: Added it.
